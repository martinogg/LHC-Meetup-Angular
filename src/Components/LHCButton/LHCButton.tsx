
import React, { Component, Children } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';

interface IProps {
    onSelected(): void
    styles?: any
}

export default class LHCButton extends Component<IProps> {

    constructor(props: IProps) {

        super(props);
    }

    render() {

        return <TouchableOpacity style={[localStyles.container, this.props.styles]} onPress={() => this.props.onSelected()}>
            {this.props.children}
        </TouchableOpacity>
    }
}

const localStyles = StyleSheet.create({
    container: {
        margin: 10,
        backgroundColor: '#2980b6',
        paddingVertical: 15
    },
})