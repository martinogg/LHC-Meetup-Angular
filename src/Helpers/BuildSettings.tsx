const jsonkey = {
    LHC: {
        FIREBASEapiKey: "FIREBASEKEY",
        FIREBASEauthDomain: "FIREBASEKEY",
        FIREBASEdatabaseURL: "FIREBASEKEY",
        FIREBASEstorageBucket: "FIREBASEKEY",
    }
};

const build = 'LHC';

class BuildSettings {
    
    public static get(value: string): any {
        const app = jsonkey[build]
        return app[value]
    }

}

export default BuildSettings;