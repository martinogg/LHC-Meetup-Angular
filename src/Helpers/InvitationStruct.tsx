export interface IInvitation {
    from: string,
    to: string,
    reason: string,
    status: InvitationStatus
}

export interface IInvitationFromFirebase {
    id: string,
    invitation: IInvitation
}

export interface IInvitationFromAndTo {
    from: IInvitationFromFirebase[],
    to: IInvitationFromFirebase[]
}

export enum InvitationStatus {
    New = 'New',
    Accepted = 'Accepted',
    Rejected = 'Rejected'
}

export class Invitation {

    public static create(_from: string, _to: string, _reason: string, _status: InvitationStatus): IInvitation {

        const ret = {
            from: _from,
            to: _to,
            reason: _reason,
            status: _status
        }
        return ret
    }
}