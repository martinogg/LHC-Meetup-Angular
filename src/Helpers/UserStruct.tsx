export interface IUserInterest {
    title: string,
    description: string
}

export interface IUser {
    userName: string,
    userLocation: string,
    userContact: string,
    userInterests: IUserInterest[],
}

export interface IUserFromFirebase {
    id: string,
    user: IUser
}

export class User {

    public static create(name: string, location: string, contact: string, interests: IUserInterest[]): IUser {

        const ret: IUser = {
            userName: name,
            userLocation: location,
            userContact: contact,
            userInterests: interests
        }

        return ret
    }
}

export class UserInterest {

    public static create(_title: string, _description: string): IUserInterest {

        const ret: IUserInterest = {
            title: _title,
            description: _description
        }

        return ret
    }
}