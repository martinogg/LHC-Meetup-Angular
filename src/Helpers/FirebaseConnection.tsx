// TODO Refactor this mess and get some tests done

'use strict'
import React from 'react';
import {
  Alert,
} from 'react-native';
import * as firebase from 'firebase';
import 'firebase/firestore';
import BuildSettings from './BuildSettings';

import { IUser, User, IUserFromFirebase, IUserInterest } from './UserStruct'
import { IInvitation, IInvitationFromFirebase, Invitation, IInvitationFromAndTo, InvitationStatus, IInvitationFromAndToWithUserObjects, IInvitationFromFirebaseWithUserObject } from './InvitationStruct'

var instance: FirebaseConnection | null = null;

class FirebaseConnection {
  firebaseApp: firebase.app.App | null = null;

  static getInstance(): FirebaseConnection {
    if (instance) {
      return instance;
    } else {
      instance = new FirebaseConnection();
      instance.initApp();
      return instance;
    }
  }

  initApp() {
    var firebaseConfig = {
      apiKey: BuildSettings.get("FIREBASEapiKey"),
      authDomain: BuildSettings.get("FIREBASEauthDomain"),
      databaseURL: BuildSettings.get("FIREBASEdatabaseURL"),
      storageBucket: BuildSettings.get("FIREBASEstorageBucket"),
      projectId: BuildSettings.get("FIREBASEprojectId")
    };

    if (firebaseConfig.apiKey != "") {
      this.firebaseApp = firebase.initializeApp(firebaseConfig);
    }
  }

  getProperties(obj, count) {
    if (isNaN(count)) count = Infinity
    var keys = []
    for (var it in obj) {
      if (keys.length > count) break;
      keys.push(it);
    }
    return keys;
  }

  public logout(): Promise<void> {

    return new Promise((resolve, reject) => {
      firebase.auth().signOut().then(() => {

        resolve()
      }, (failReason) => {

        Alert.alert('Error: ' + failReason)
        resolve()
      })
    })
  }

  // TODO TEST
  public checkFBVersion(): Promise<boolean> {
    return new Promise<boolean>((resolve, reject) => {

      if (firebase.auth().currentUser == null) {

        reject('User Not Logged in')
        return
      }

      const localVersion: number = BuildSettings.get("FIREBASEVersion")

      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-settings").get()
        .then((snap) => {

          let fbVersion: number = 99999999
          snap.forEach(setting => {

            if (setting.id == 'version') {

              const data = setting.data()
              fbVersion = +data.version
            }

          });

          resolve(localVersion >= fbVersion)
        })
        .catch(function (error) {

          reject(error)
        });
    })
}


  //TODO TEST
  public buildInvitationsWithUserObjects(invitations: IInvitationFromAndTo): Promise < IInvitationFromAndToWithUserObjects > {

  return new Promise<IInvitationFromAndToWithUserObjects>((resolve, reject) => {

    this.allUsers().then((users) => {

      let fromInvitations: IInvitationFromFirebaseWithUserObject[] = []
      let toInvitations: IInvitationFromFirebaseWithUserObject[] = []

      const createNewInvitation: (invitation: IInvitationFromFirebase) => IInvitationFromFirebaseWithUserObject = (invitation) => {

        const newInvitation: IInvitationFromFirebaseWithUserObject = {

          ...invitation,
          fromObject: this.userObjectForID(invitation.invitation.from, users),
          toObject: this.userObjectForID(invitation.invitation.to, users)
        }

        return newInvitation
      }

      invitations.from.forEach((element) => {

        fromInvitations.push(createNewInvitation(element))
      })

      invitations.to.forEach((element) => {

        toInvitations.push(createNewInvitation(element))
      })

      resolve({ from: fromInvitations, to: toInvitations })

    }).catch((error) => {

      reject(error)
    })

  })
}

  private userObjectForID(id: string, users: IUserFromFirebase[]): IUserFromFirebase {

  let ret: IUserFromFirebase | null = null
  users.forEach((element) => {

    if (element.id == id) {

      ret = element
    }
  })

  if (ret == null) {

    ret = { id: id, user: User.create('Unknown User', '', '', []) }
  }

  return ret
}

  public getInvitations(): Promise < IInvitationFromAndTo > {
  // TODO TEST
  return new Promise<IInvitationFromAndTo>((resolve, reject) => {

    if (firebase.auth().currentUser == null) {

      reject('User Not Logged in')
      return
    }

    const currentUser = firebase.auth().currentUser as firebase.User

    firebase.firestore().collection("LHC-Invitations").get()
      .then((snap) => {

        let ret: IInvitationFromFirebase[] = []

        snap.forEach(invitationInList => {

          const id = invitationInList.id
          const data = invitationInList.data()

          if (data) {
            
            const invitation = Invitation.create(data.from, data.to, data.reason, data.status)
            ret.push({ id: id, invitation: invitation })
          }

        });

        resolve({ from: ret.filter((item) => { return item.invitation.from == currentUser.uid }), to: ret.filter((item) => { return item.invitation.to == currentUser.uid }) })
      })
      .catch(function (error) {

        reject(error)
      });
  })

}

  public deleteInvitation(uid: string): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      firebase.firestore().collection("LHC-Invitations").doc(uid).delete()
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}


  public updateInvitationResponse(uid: string, response: InvitationStatus): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      firebase.firestore().collection("LHC-Invitations").doc(uid).update({ status: response })
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  public updateInvitation(uid: string, reason: string): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      firebase.firestore().collection("LHC-Invitations").doc(uid).update({ reason })
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}


  public createNewInvitation(invitation: IInvitation): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      const newInvitationId = this.dateEpochPlusUID()
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Invitations").doc(newInvitationId).set(invitation)
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  public saveUserDetails(user: IUser): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").doc(currentUser.uid).set({
        name: user.userName,
        location: user.userLocation,
        contact: user.userContact,
        interests: user.userInterests,
      })
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  private queryMatchesInterest(query: string, interestList: IUserInterest[]): boolean {

  let found = false
  const queryLower = query.toLowerCase()
  interestList.forEach((interest) => {

    if (interest.title.toLowerCase().includes(queryLower)) {

      found = true
    }
    if (interest.description.toLowerCase().includes(queryLower)) {

      found = true
    }
  })

  return found
}

  public searchOtherUsers(query: string): Promise < IUserFromFirebase[] > {
  // TODO TEST
  return new Promise<IUserFromFirebase[]>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").get()
        .then((snap) => {

          let ret: IUserFromFirebase[] = []

          snap.forEach(userInList => {

            const id = userInList.id
            const data = userInList.data()
            const matchesInterest = this.queryMatchesInterest(query, data.interests)

            if (data != null && id != currentUser.uid && matchesInterest) {

              const user = User.create(data.name, data.location, data.contact, data.interests)
              ret.push({ id: id, user: user })
            }

          });

          resolve(ret)
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })

}

  public allUsers(): Promise < IUserFromFirebase[] > {
  // TODO TEST
  return new Promise<IUserFromFirebase[]>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").get()
        .then((snap) => {

          let ret: IUserFromFirebase[] = []

          snap.forEach(userInList => {

            const id = userInList.id
            const data = userInList.data()

            const user = User.create(data.name, data.location, data.contact, data.interests)
            ret.push({ id: id, user: user })

          });

          resolve(ret)
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })

}

  public loadUserDetails(): Promise < IUser > {
  // TODO TEST
  return new Promise<IUser>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").doc(currentUser.uid).get()
        .then((snap) => {

          const data = snap.data()

          if (data != null) {
            const user = User.create(data.name, data.location, data.contact, data.interests)

            resolve(user)
          } else {
            reject('An Error Occurred')
          }
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  public login(username: string, password: string): Promise < string > {
  // TODO TEST

  return new Promise((resolve, reject) => {

    firebase.auth().signInWithEmailAndPassword(username, password).then(() => {

      resolve()
    }, (reason) => {

      reject(reason)
    })
  })
}

  public register(username: string, password: string): Promise < string > {
  // TODO TEST

  return new Promise((resolve, reject) => {

    firebase.auth().createUserWithEmailAndPassword(username, password).then(() => {


      // TODO - Shift this out in to separate email verification
      /*
      if (this.isLoggedIn()) {
 
        firebase.auth().currentUser.sendEmailVerification()
      }
      */

      resolve()
    }, (reason) => {
      reject(reason)
    })
  })
}

  public isLoggedIn(): boolean {
  // TODO TEST

  return firebase.auth().currentUser != null
}

  public getCurrentUserID(): string {
  // TODO TEST
  if (firebase.auth().currentUser == null) {
    return 'Error: Not logged In'
  }

  const currentUser = firebase.auth().currentUser as firebase.User
  return currentUser.uid
}

  private dateEpochPlusUID(): string {
  // TODO TEST
  return (new Date).getTime().toString() + '-' + (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1)
}

sendComment(comment: string) {
  // TODO TEST

  if (comment !== '') {

    const date = this.dateEpochPlusUID()
    if (firebase.auth().currentUser != null) {

      const currentUser = firebase.auth().currentUser as firebase.User
      firebase.firestore().collection("LHC-Comments").doc(date).set({
        'comment': comment
      })
    }
  }
}

loginAnon() {
  let strongThis = this;
  if (this.firebaseApp != null) {
    firebase.auth().signInAnonymously().catch((error) => { });
  }
}

// only allows callback if user is actually logged in
_getCurrentUser(onGetCurrentUserFunc) {

  if (this.firebaseApp != null) {

    var currentUser = firebase.auth().currentUser;
    if (currentUser) {

      onGetCurrentUserFunc(currentUser);
    } else {

      // try again just in case user is back online
      this.loginAnon();
    }
  }
}

getLikeWithId(idStr, callback) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("likes").child(idStr);
    ref.once('value', (snap) => {
      var isSet = snap.val() === true;
      callback(isSet);
    });
  });
}

toggleLikeWithId(idStr, callback) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("likes").child(idStr);

    ref.once('value', (snap) => {
      var alreadySet = snap.val() === true;
      var newSet = alreadySet ? null : true;
      ref.set(newSet).then(() => {
        callback(newSet);
      });
    });
  });
}

setViewedWithId(idStr, onCompleteFunc) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("views").child(idStr);

    ref.set(true).then(() => {
      onCompleteFunc();
    });
  });
}

setLikeWithId(idStr, liked, onCompleteFunc) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("likes").child(idStr);

    ref.set(liked ? true : null).then(() => {
      onCompleteFunc();
    });
  });
}

setDownloadedWithId(idStr) {
  this._getCurrentUser((currentUser) => {
    this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("downloads").child(idStr).set(true);
  });
}

getDbDecks(onCompleteFunc) {

  if (this.firebaseApp != null) {

    const strongThis = this;

    var itemsRef = this.firebaseApp.database().ref().child("data");

    itemsRef.once('value', (snap) => {

      // get children as an array
      var items = [];
      snap.forEach((child) => {
        items.push(child);
      });
      onCompleteFunc(items);
    });
  }
}

getDbDictLikes(onCompleteFunc) {

  if (this.firebaseApp != null) {
    const strongThis = this;

    var itemsRef = this.firebaseApp.database().ref().child("data");

    itemsRef.once('value', (snap) => {

      // get children as an array
      var items = {};
      snap.forEach((child) => {
        var likesCount = 0 + child.val().likes;
        var viewsCount = 0 + child.val().views;
        items["" + child.key] = { likes: likesCount, views: viewsCount };
      });
      onCompleteFunc(items);
    });
  }
}



alert(message) {
  Alert.alert(
    'Error',
    'Error:' + message,
    [
      { text: 'OK', onPress: () => console.log('OK Pressed') },
    ],
    { cancelable: false }
  );
}

}

export default FirebaseConnection // TODO FirebaseConnection is old style, remove this in favor of REDUX Version in the future

export const firebaseService = FirebaseConnection.getInstance().firebaseApp as firebase.app.App