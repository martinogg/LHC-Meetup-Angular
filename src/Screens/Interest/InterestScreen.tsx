import React, { Component } from 'react';
import { Keyboard, KeyboardAvoidingView, StyleSheet, TextInput, Text, View, Alert, SafeAreaView, FlatList } from 'react-native';
import { NavigationActions, NavigationScreenProp, StackActions } from 'react-navigation'
import FirebaseConnection from '../../Helpers/FirebaseConnection'
import { IUser, User, IUserInterest, UserInterest } from '../../Helpers/UserStruct'
import LHCButton from '../../Components/LHCButton/LHCButton'
import { AppStyles } from '../../AppStyles'

interface IProps {
    navigation: NavigationScreenProp<any, any>,
}

interface IState {
    userInterest: IUserInterest,
    title: string,
    description: string
}

export class InterestScreen extends Component<IProps, IState> {

    private static navigationOptions = {
        title: 'Interest'
    };

    constructor(props: IProps) {

        super(props);
        const userInterest = this.props.navigation.getParam('previousUserInterest')
        this.state = {
            userInterest: userInterest,
            title: userInterest.title,
            description: userInterest.description
        };
    }

    saveButtonPressed() {

        const saveCallback: (newInterest: UserInterest) => void = this.props.navigation.getParam('saveCallback')
        const newInterest = UserInterest.create(this.state.title, this.state.description)
        saveCallback(newInterest)
        this.props.navigation.pop()
    }

    removeButtonPressed() {

        const removeCallback = this.props.navigation.getParam('removeCallback')
        removeCallback()
        this.props.navigation.pop()
    }

    public render() {

        return (
            <SafeAreaView style={AppStyles.container}>

                <TextInput style={AppStyles.input}
                    autoCapitalize="none"
                    onSubmitEditing={() => this.contactInput.focus()}
                    autoCorrect={false}
                    keyboardType='default'
                    returnKeyType="next"
                    placeholder='Interest Title'
                    value={this.state.title}
                    placeholderTextColor='rgba(225,225,225,0.7)'
                    onChangeText={(text) => this.handleTitleChange(text)}
                />

                <TextInput style={AppStyles.input}
                    ref={(contact) => this.contactInput = contact}
                    autoCapitalize="none"
                    autoCorrect={false}
                    keyboardType='default'
                    returnKeyType="done"
                    value={this.state.description}
                    placeholder='Description'
                    placeholderTextColor='rgba(225,225,225,0.7)'
                    onChangeText={(text) => this.handleDescriptionChange(text)}
                />

                <View style={{ flex: 1 }} />

                <LHCButton onSelected={() => this.saveButtonPressed()}>
                    <Text style={AppStyles.buttonText}>Save Changes</Text>
                </LHCButton>

                <LHCButton onSelected={() => this.removeButtonPressed()}>
                    <Text style={AppStyles.buttonText}>Remove interest</Text>
                </LHCButton>

            </SafeAreaView>
        );
    }

    private contactInput: any

    private handleTitleChange(text: string) {

        this.setState({ title: text })
    }

    private handleDescriptionChange(text: string) {

        this.setState({ description: text })
    }
}

const styles = StyleSheet.create({
    entriesContainer: {
        flex: 1,
        padding: 20
    },
    loginContainer: {
        alignItems: 'center',
        justifyContent: 'center'
    },
    logo: {
        position: 'absolute',
        width: 300,
        height: 100
    },
    title: {
        color: "#FFF",
        marginTop: 120,
        width: 180,
        textAlign: 'center',
        opacity: 0.9
    }
})