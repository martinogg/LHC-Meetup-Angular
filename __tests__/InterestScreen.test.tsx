import { shallow } from 'enzyme';
import React from 'react';
import 'react-native';
import { Alert } from 'react-native'
import { User, IUser, UserInterest } from '../src/Helpers/UserStruct'
import LHCButton from '../src/Components/LHCButton/LHCButton'
import { InterestScreen } from '../src/Screens/Interest/InterestScreen';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const createTestProps = (props: Object) => ({
    navigation: {
        navigate: jest.fn(),
        replace: jest.fn(),
        dispatch: jest.fn(),
        getParam: (param: string) => { return {} },
        state: {
            params: {
                editable: true
            }
        }
    },
    ...props
});

it('should display InterestScreen Editable with no errors', () => {

    let props = createTestProps({
        previousUserInterest: UserInterest.create('a', 'b'),
        screenProp: {}
    });

    const newNavigation = {
        ...props.navigation,
        getParam: (param: string) => { return {} },
    }
    props.navigation = newNavigation

    expect(renderer.create(<InterestScreen {...props} />)).toMatchSnapshot();
});

it('should display InterestScreen Non-Editable with no errors', () => {

    let props = createTestProps({
        previousUserInterest: UserInterest.create('a', 'b'),
        screenProp: {}
    });

    const newNavigation = {
        ...props.navigation,
        getParam: (param: string) => { return {} },
        state: {
            params: {
                editable: false
            }
        }
    }
    props.navigation = newNavigation

    expect(renderer.create(<InterestScreen {...props} />)).toMatchSnapshot();
});

test('InterestScreen constructor ', () => {

    const mockInterest = UserInterest.create('a', 'b')
    let props = createTestProps({
        previousUserInterest: UserInterest.create('a', 'b'),
        screenProp: {}
    });
    const newNavigation = {
        ...props.navigation,
        getParam: (param: string) => { return mockInterest },
    }
    props.navigation = newNavigation

    const sut: any = new InterestScreen(props)

    expect(sut.state.title).toBe(mockInterest.title)
    expect(sut.state.description).toBe(mockInterest.description)
    expect(sut.state.userInterest).toBe(mockInterest)
});

test('InterestScreen saveButtonPressed function', () => {

    const mockCallbackFunc = jest.fn()
    let getParamCallback: (param: string) => any = (param) => { return {} }
    const mockInterest = UserInterest.create('a', 'b')
    let props = createTestProps({
        
        previousUserInterest: UserInterest.create('a', 'b'),
        screenProp: { }
    });
    const newNavigation = {
        ...props.navigation,
        getParam: (param: string) => { return getParamCallback(param) },
        pop: jest.fn()
    }
    props.navigation = newNavigation

    const sut: any = new InterestScreen(props)
    getParamCallback = (param: string) => {
        expect(param).toEqual('saveCallback')
        return mockCallbackFunc
    }

    sut.state.title = mockInterest.title
    sut.state.description = mockInterest.description

    sut.saveButtonPressed()

    expect(mockCallbackFunc).toHaveBeenCalledTimes(1)
    expect(mockCallbackFunc).toHaveBeenCalledWith(expect.objectContaining(mockInterest))
    expect(props.navigation.pop).toHaveBeenCalledTimes(1)
})

test('InterestScreen removeButtonPressed function', () => {

    const mockCallbackFunc = jest.fn()
    let getParamCallback: (param: string) => any = (param) => { return {} }
    const mockInterest = UserInterest.create('a', 'b')
    const props = createTestProps({
        navigation: {
            getParam: (param: string) => { return getParamCallback(param) },
            pop: jest.fn()
        },
        previousUserInterest: UserInterest.create('a', 'b'),
        screenProp: {

        }
    });

    const sut: any = new InterestScreen(props)

    getParamCallback = (param: string) => {
        expect(param).toEqual('removeCallback')
        return mockCallbackFunc
    }

    sut.state.title = mockInterest.title
    sut.state.description = mockInterest.description

    sut.removeButtonPressed()

    expect(mockCallbackFunc).toHaveBeenCalledTimes(1)
    expect(props.navigation.pop).toHaveBeenCalledTimes(1)
})

test('InterestScreen handleTitleChange function', () => {

    const props = createTestProps({});
    const changeText = 'change'

    const wrapper = shallow(<InterestScreen {...props} />);
    const sut: any = wrapper.instance()
    sut.handleTitleChange(changeText)
    expect(sut.state.title).toEqual(changeText)
})

test('InterestScreen handleDescriptionChange function', () => {

    const props = createTestProps({});
    const changeText = 'change'

    const wrapper = shallow(<InterestScreen {...props} />);
    const sut: any = wrapper.instance()
    sut.handleDescriptionChange(changeText)
    expect(sut.state.description).toEqual(changeText)
})
