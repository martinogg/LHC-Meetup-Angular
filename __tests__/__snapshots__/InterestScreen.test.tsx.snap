// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display InterestScreen Editable with no errors 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#2c3e50",
      "flex": 1,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    autoCorrect={false}
    editable={true}
    keyboardType="default"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Interest Title"
    placeholderTextColor="rgba(225,225,225,0.7)"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Object {
        "backgroundColor": "rgba(225,225,225,0.2)",
        "color": "#fff",
        "height": 40,
        "marginBottom": 10,
        "padding": 10,
      }
    }
    underlineColorAndroid="transparent"
  />
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    autoCorrect={false}
    editable={true}
    keyboardType="default"
    onChangeText={[Function]}
    placeholder="Description"
    placeholderTextColor="rgba(225,225,225,0.7)"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Object {
        "backgroundColor": "rgba(225,225,225,0.2)",
        "color": "#fff",
        "height": 40,
        "marginBottom": 10,
        "padding": 10,
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#2980b6",
          "margin": 10,
          "opacity": 1,
          "paddingVertical": 15,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
      >
        Save Changes
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#2980b6",
          "margin": 10,
          "opacity": 1,
          "paddingVertical": 15,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
      >
        Remove interest
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`should display InterestScreen Non-Editable with no errors 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#2c3e50",
      "flex": 1,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    autoCorrect={false}
    editable={false}
    keyboardType="default"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Interest Title"
    placeholderTextColor="rgba(225,225,225,0.7)"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Object {
        "backgroundColor": "rgba(225,225,225,0.2)",
        "color": "#fff",
        "height": 40,
        "marginBottom": 10,
        "padding": 10,
      }
    }
    underlineColorAndroid="transparent"
  />
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    autoCorrect={false}
    editable={false}
    keyboardType="default"
    onChangeText={[Function]}
    placeholder="Description"
    placeholderTextColor="rgba(225,225,225,0.7)"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Object {
        "backgroundColor": "rgba(225,225,225,0.2)",
        "color": "#fff",
        "height": 40,
        "marginBottom": 10,
        "padding": 10,
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</RCTSafeAreaView>
`;
